{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "def transcription():\n",
    "    mRNA = []\n",
    "    for i in range(0,len(code)):\n",
    "        if code[i] == 'A':\n",
    "            mRNA.append('U')\n",
    "        elif code[i] == 'T':\n",
    "            mRNA.append('A')\n",
    "        elif code[i] == 'C':\n",
    "            mRNA.append('G')\n",
    "        elif code[i] == 'G':\n",
    "            mRNA.append('C')\n",
    "    global mRNA_code\n",
    "    mRNA_code = ''.join(mRNA)\n",
    "\n",
    "def translation():   \n",
    "    global anti_codons \n",
    "    anti_codons = []\n",
    "    for i in range(0,len(mRNA_code)):\n",
    "        if mRNA_code[i] == 'A':\n",
    "            anti_codons.append('U')\n",
    "        elif mRNA_code[i] == 'U':\n",
    "            anti_codons.append('A')\n",
    "        elif mRNA_code[i] == 'C':\n",
    "            anti_codons.append('G')\n",
    "        elif mRNA_code[i] == 'G':\n",
    "            anti_codons.append('C')\n",
    "    global anti_codons_code\n",
    "    anti_codons_code = ''.join(anti_codons)\n",
    "\n",
    "def polypeptide_chain():\n",
    "    codon_dict = {\n",
    "        'UUU':'Phe','UUC':'Phe','UUA':'Leu','UUG':'Leu',\n",
    "        'UCU':'Ser','UCC':'Ser','UCA':'Ser','UCG':'Ser',\n",
    "        'UAU':'Tyr','UAC':'Tyr','UAA':'STOP','UAG':'STOP',\n",
    "        'UGU':'Cyr','UGC':'Cyr','UGA':'STOP','UGG':'Trp',\n",
    "        'CUU':'Leu','CUC':'Leu','CUA':'Leu','CUG':'Leu',\n",
    "        'CCU':'Pro','CCC':'Pro','CCA':'Pro','CCG':'Pro',\n",
    "        'CAU':'His','CAC':'His','CAA':'Gln','CAG':'Gln',\n",
    "        'CGU':'Arg','CGC':'Arg','CGA':'Arg','CGG':'Arg',\n",
    "        'AUU':'Ile','AUC':'Ile','AUA':'Ile','AUG':'Met',\n",
    "        'ACU':'Thr','ACC':'Thr','ACA':'Thr','ACG':'Thr',\n",
    "        'AAU':'Asn','AAC':'Asn','AAA':'Lys','AAG':'Lyr',\n",
    "        'AGU':'Ser','AGC':'Ser','AGA':'Arg','AGG':'Arg',\n",
    "        'GUU':'Val','GUC':'Val','GUA':'Val','GUG':'Val',\n",
    "        'GCU':'Ala','GCC':'Ala','GCA':'Ala','GCG':'Ala',\n",
    "        'GAU':'Asp','GAC':'Asp','GAA':'Glu','GAG':'Glu',\n",
    "        'GGU':'Gly','GGC':'Gly','GGA':'Gly','GGG':'Gly'}\n",
    "    \n",
    "    \n",
    "    anti_codon_triplets = [anti_codons_code[i:i+3] for i in range(0, len(anti_codons_code), 3)]\n",
    "    if (anti_codon_triplets[0] == 'UGA') or (anti_codon_triplets[0] == 'UAG') or (anti_codon_triplets[0] == 'UAA'):\n",
    "        print('This is an empty polypeptide chain.')\n",
    "        \n",
    "    if len(anti_codon_triplets[len(anti_codon_triplets)-1]) < 3:\n",
    "        anti_codon_triplets.remove(anti_codon_triplets[len(anti_codon_triplets)-1])\n",
    "       \n",
    "    if len(anti_codon_triplets[len(anti_codon_triplets)-1]) == 1:\n",
    "        add_info = '*There was an extra nitrogen base in your DNA strand, so we removed it from the DNA transcript during transcription.'\n",
    "    else:\n",
    "        add_info = '*There were extra nitrogen bases in your DNA strand, so we removed them from the DNA transcript during transcription.'\n",
    "   \n",
    "    polypeptide_chain = []\n",
    "    for codon in anti_codon_triplets:\n",
    "        if (codon_dict[codon] == 'STOP'):\n",
    "            break\n",
    "        polypeptide_chain.append(codon_dict[codon])\n",
    "    print('> '+'-'.join(polypeptide_chain))\n",
    "    if len(code) % 3 != 0:\n",
    "        print(add_info)\n",
    "\n",
    "def protein_synthesis():\n",
    "    global code \n",
    "    code = input('Please enter a valid strand of DNA: (containing A,T,G,or C only)').upper().replace(' ','').replace('-','')\n",
    "    DNA_nucleotides = {'A','C','G','T'}\n",
    "    \n",
    "    if (set(sorted(set(code))).issubset(DNA_nucleotides)):\n",
    "        print()\n",
    "        print('Commencing Protein Synthesis...')\n",
    "        time.sleep(3)\n",
    "        print()\n",
    "        print('- Transcription ✓')\n",
    "        time.sleep(3)\n",
    "        print()\n",
    "        print('- Translation ✓')\n",
    "        time.sleep(3)\n",
    "        print()\n",
    "        print('- Protein Synthesis ✓')\n",
    "        time.sleep(3)\n",
    "        print()\n",
    "        print('Here is the resulting protein: ')\n",
    "        time.sleep(5)\n",
    "        print()\n",
    "        print()\n",
    "        transcription()\n",
    "        translation()\n",
    "        polypeptide_chain()\n",
    "    \n",
    "    elif 'U' in code:\n",
    "        print('Loading...')\n",
    "        time.sleep(3)\n",
    "        print()\n",
    "        print('Uracil found: Please enter a DNA strand not an RNA strand.')\n",
    "        protein_synthesis()\n",
    "    \n",
    "    else:\n",
    "        print('Loading...')\n",
    "        time.sleep(3)\n",
    "        print()\n",
    "        print('This is an invalid DNA sequence. Please enter a correct sequence: ')\n",
    "        protein_synthesis()\n",
    "        \n",
    "protein_synthesis()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
